shared/mad/codeql/mad/ModelValidation.qll

```diff
@@ -0,0 +1,177 @@
+          "%string-%length", "weak-hash-input-%"
+        ])
+    }
+  }
+
+  /** An outdated models-as-data sink kind. */
+  private class OutdatedSinkKind extends string {
+    OutdatedSinkKind() {
+      this =
+        [
+          "sql", "url-redirect", "xpath", "ssti", "logging", "groovy", "jexl", "mvel", "xslt",
+          "ldap", "pending-intent-sent", "intent-start", "set-hostname-verifier",
+          "header-splitting", "xss", "write-file", "create-file", "read-file", "open-url",
+          "jdbc-url", "command-line-injection", "code", "html", "remote",
+          "uncontrolled-format-string", "js-eval"
+        ]
+    }
+
+    /** Gets a replacement kind for an outdated sink kind. */
+    private string replacementKind() {
+      this = ["sql", "xpath", "groovy", "jexl", "mvel", "xslt", "ldap", "code", "html"] and
+      result = this + "-injection"
+      or
+      this = "js-eval" and result = "code-injection"
+      or
+      this = "url-redirect" and result = "url-redirection"
+      or
+      this = "ssti" and result = "template-injection"
+      or
+      this = "logging" and result = "log-injection"
+      or
+      this = "pending-intent-sent" and result = "pending-intents"
+      or
+      this = "intent-start" and result = "intent-redirection"
+      or
+      this = "set-hostname-verifier" and result = "hostname-verification"
+      or
+      this = "header-splitting" and result = "response-splitting"
+      or
+      this = "xss" and result = "html-injection\" or \"js-injection"
+      or
+      this = ["write-file", "remote"] and result = "file-content-store"
+      or
+      this = ["create-file", "read-file"] and result = "path-injection"
+      or
+      this = ["open-url", "jdbc-url"] and result = "request-forgery"
+      or
+      this = "command-line-injection" and result = "command-injection"
+      or
+      this = "uncontrolled-format-string" and result = "format-string"
+    }
+
+    /** Gets an error message for an outdated sink kind. */
+    string outdatedMessage() {
+      result =
+        "The kind \"" + this + "\" is outdated. Use \"" + this.replacementKind() + "\" instead."
+    }
+  }
+
+  /** A valid models-as-data source kind. */
+  private class ValidSourceKind extends string {
+    bindingset[this]
+    ValidSourceKind() {
+      this =
+        [
+          // shared
+          "local", "remote",
+          // Java
+          "android-external-storage-dir", "contentprovider",
+          // C#
+          "file", "file-write",
+          // JavaScript
+          "database-access-result"
+        ]
+      or
+      this.matches([
+          // shared
+          "qltest%", "test-%",
+          // Swift
+          "%string-%length"
+        ])
+    }
+  }
+
+  /** A valid models-as-data summary kind. */
+  private class ValidSummaryKind extends string {
+    ValidSummaryKind() {
+      this =
+        [
+          // shared
+          "taint", "value",
+          // Dynamic languages
+          "type"
```

Comment:
I think that for all practical purposes for now, it is acceptable the shared code also contain some validation that is language specific (to reduce the amount of code we need to write and maintain). If we notice (going forward) that the we have to much language specific stuff in the shared validation we probably need to split it and make sure that the shared validation is only the intersection between the languages.